# Algoritmos de optimización - Seminario<br>
Nombre y Apellidos:   <br>
* Alberto Paparelli<br>
* Raúl Sánchez Mondéjar

Url: https://github.com/carpe-diem/VIU-03MIAR-algoritmos-de-optimizacion/blob/main/Proyecto/Seminario_Algoritmos.ipynb<br>

Problema:
> 2. Organizar los horarios de partidos de La Liga<br>

Descripción del problema:

* Desde la La Liga de fútbol profesional se pretende organizar los horarios de los partidos de liga de cada jornada. Se conocen algunos datos que nos deben llevar a diseñar un algoritmo que realice la asignación de los partidos a los horarios de forma que maximice la audiencia.
* Los horarios disponibles se conocen a priori y son los siguientes:

| Viernes 	| 20             	|
|---------	|----------------	|
| Sábado  	| 12, 16, 18, 20 	|
| Domingo 	| 12, 16, 18, 20 	|
| Lunes   	| 20             	|

* En primer lugar se clasifican los equipos en tres categorías según el numero de seguidores( que tiene relación directa con la audiencia). Hay 3 equipos en la categoría A, 11 equipos de categoría B y 6 equipos de categoría C.
* Se conoce estadísticamente la audiencia que genera cada partido según los equipos que se enfrentan y en horario de sábado a las 20h (el mejor en todos los casos).

|             	| Categoría A 	|  Categoría B 	|  Categoría C  	|
|-------------	|:-----------:	|:------------:	|:-------------:	|
| Categoría A 	|  2 Millones 	| 1.3 Millones 	|    1 Millon   	|
| Categoría B 	|             	| 0.9 Millones 	| 0.75 Millones 	|

* Si el horario del partido no se realiza a las 20 horas del sábado se sabe que se reduce según los coeficientes de la siguiente tabla.
* Debemos asignar obligatoriamente siempre un partido el viernes y un partido el lunes.

|     	| Viernes 	| Sábado 	| Domingo 	| Lunes 	|
|-----	|---------	|--------	|---------	|-------	|
| 12h 	| -       	| 0.55   	| 0.45    	| -     	|
| 16h 	| -       	| 0.7    	| 0.75    	| -     	|
| 18h 	| -       	| 0.8    	| 0.85    	| -     	|
| 20h 	| 0.4     	| 1      	| 1       	| 0.4   	|

* Es posible la coincidencia de horarios pero en este caso la audiencia de cada partido se verá afectada y se estima que se reduce en porcentaje según la siguiente tabla dependiendo del número de coincidencias:

| Coincidencias 	|  -% 	|
|:-------------:	|:---:	|
|       0       	|  0% 	|
|       1       	| 25% 	|
|       2       	| 45% 	|
|       3       	| 60% 	|
|       4       	| 70% 	|
|       5       	| 75% 	|
|       6       	| 78% 	|
|       7       	| 80% 	|
|       8       	| 80% 	|

* Los cálculos asociados a una jornada de ejemplo se realizan según se muestra en la siguiente tabla:

|          Partido         	| Categorías 	| Horario 	| Base (Millones) 	| Ponderación 	| Base * Ponderación 	| Corrección Coincidencia 	|
|:------------------------:	|:----------:	|:-------:	|:---------------:	|:-----------:	|:------------------:	|:-----------------------:	|
| Celta      - Real Madrid 	|     B-A    	|   V20   	|       1.3       	|     0.4     	|        0.52        	|           0.52          	|
| Valencia   - R. Sociedad 	|     B-A    	|   S12   	|       1.3       	|     0.55    	|        0.72        	|           0.72          	|
| Mallorca   - Eibar       	|     C-C    	|   S16   	|       0.47      	|     0.7     	|        0.33        	|           0.33          	|
| Athletic   - Barcelona   	|     B-A    	|   S18   	|       1.3       	|     0.8     	|        1.04        	|           1.04          	|
| Leganés    - Osasuna     	|     C-C    	|   S20   	|       0.47      	|      1      	|        0.47        	|           0.47          	|
| Villarreal - Granada     	|   **B-C**  	| **D16** 	|     **0.75**    	|   **0.75**  	|      **0.56**      	|         **0.42** *(1)    	|
| Alavés     - Levante     	|   **B-B**  	| **D16** 	|     **0.9**     	|   **0.75**  	|      **0.68**      	|         **0.51** *(2)    	|
| Espanyol   - Sevilla     	|     B-B    	|   D18   	|       0.9       	|     0.85    	|        0.77        	|           0.77          	|
| Betis      - Valladolid  	|     B-C    	|   D20   	|       0.75      	|      1      	|        0.75        	|           0.75          	|
| Atlético   - Getafe      	|     B-B    	|   L20   	|       0.9       	|     0.4     	|        0.36        	|           0.36          	|
|||||||                                                                                                                     TOTAL: 5.88         |

*(1) 0.42 = 0.56 x 0.75

*(2) 0.51 = 0.68 x 0.75


(*) La respuesta es obligatoria

---

### * ¿Cuantas posibilidades hay sin tener en cuenta las restricciones?

Datos:
- 10 partidos por jornada (20 equipos / 2)
- 10 franjas horarias disponibles: V20, S12, S16, S18, S20, D12, D16, D18, D20, L20

Sin restricciones, cada partido puede asignarse a cualquiera de las 10 franjas horarias,
y se permite que varios partidos coincidan en la misma franja.

Esto es un problema de ASIGNACIÓN CON REPETICIÓN: a cada uno de los n partidos
le asignamos una de las k franjas horarias posibles.

Por el principio multiplicativo:
  - Partido 1 -> 10 opciones
  - Partido 2 -> 10 opciones
  - ...
  - Partido 10 -> 10 opciones

Total = k^n = 10^10

In [2]:
n_partidos = 10   # partidos por jornada
k_horarios = 10   # franjas horarias disponibles

# Total de asignaciones posibles (asignación con repetición)
total_posibilidades = k_horarios ** n_partidos

print(f"Partidos por jornada: {n_partidos}")
print(f"Franjas horarias disponibles: {k_horarios}")
print(f"Total de posibilidades (sin restricciones): {k_horarios}^{n_partidos} = {total_posibilidades:,}")


Partidos por jornada: 10
Franjas horarias disponibles: 10
Total de posibilidades (sin restricciones): 10^10 = 10,000,000,000


### Respuesta:

Total de posibilidades (sin restricciones):
$$10^{10} = 10,000,000,000$$

---

### ¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones.

Restricciones:
    * Obligatoriamente debe haber al menos 1 partido el viernes.
    * Obligatoriamente debe haber al menos 1 partido el lunes.

Para contar las asignaciones válidas usamos el principio de inclusión-exclusión:

Sea:
- $T$ = total sin restricciones = $10^{10}$
- $A$ = asignaciones donde **ningún** partido se juega el viernes → cada partido elige entre 9 slots = $9^{10}$
- $B$ = asignaciones donde **ningún** partido se juega el lunes → $9^{10}$
- $A \cap B$ = asignaciones sin viernes **ni** lunes → cada partido elige entre 8 slots = $8^{10}$

Aplicando inclusión-exclusión:

$$\text{Válidas} = T - |A| - |B| + |A \cap B| = 10^{10} - 2 \times 9^{10} + 8^{10}$$

In [9]:
total = 10**10           # sin restricciones
sin_viernes = 9**10      # ningún partido el viernes (9 slots disponibles)
sin_lunes = 9**10        # ningún partido el lunes (9 slots disponibles)
sin_ambos = 8**10        # ningún partido el viernes ni el lunes (8 slots)

# Inclusión-exclusión: al menos 1 viernes Y al menos 1 lunes
validas = total - sin_viernes - sin_lunes + sin_ambos

print(f"Total sin restricciones:          {total:>15,}")
print(f"Sin viernes (9^10):               {sin_viernes:>15,}")
print(f"Sin lunes (9^10):                 {sin_lunes:>15,}")
print(f"Sin viernes ni lunes (8^10):      {sin_ambos:>15,}")
print(f"Total con restricciones:          {validas:>15,}")

Total sin restricciones:           10,000,000,000
Sin viernes (9^10):                 3,486,784,401
Sin lunes (9^10):                   3,486,784,401
Sin viernes ni lunes (8^10):        1,073,741,824
Total con restricciones:            4,100,173,022


### Respuesta:

Teniendo en cuenta la restricción de al menos 1 partido el viernes y 1 el lunes, el número de asignaciones válidas es:

$$10^{10} - 2 \times 9^{10} + 8^{10} = 4,100,173,022$$

---

Modelo para el espacio de soluciones<br>

### (*) ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.<br> (Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguementalo)


### Respuesta: Vector (lista Python)

**Representación:**

Utilizamos una **lista de 10 enteros** donde la posición `i` representa el partido `i` y el valor almacenado indica el índice de la franja horaria asignada (0-9).

```
solucion = [3, 0, 5, 7, 4, 6, 6, 8, 9, 1]
             ^  ^  ^  ...
          P0->S18  P1->V20  P2->D12 ...
```

**Mapeo de franjas horarias:**

| Índice | Franja |
|:------:|:------:|
| 0 | V20 |
| 1 | S12 |
| 2 | S16 |
| 3 | S18 |
| 4 | S20 |
| 5 | D12 |
| 6 | D16 |
| 7 | D18 |
| 8 | D20 |
| 9 | L20 |

**Argumentación:**

1. **Consistencia con lo visto en clase:** Es la misma estructura utilizada en el problema de N-Reinas (AG1): un vector donde la posición representa la entidad a asignar y el valor representa la asignación. Además, el manual y las actividades guiadas usan vectores para problemas de asignación.

2. **Eficiencia computacional:** Acceso O(1) por índice, tamaño fijo conocido (n=10), fácil generación de vecinos (cambiar un valor en una posición), y compatible con fuerza bruta (recorrer las $10^{10}$ combinaciones mediante conteo en base 10).

3. **Verificación de restricciones simple:** Contar coincidencias por franja con un conteo manual mediante un vector auxiliar de tamaño k (O(n)), verificar la presencia de viernes (valor 0) o lunes (valor 9) con una simple comparación de valores en el vector.

**Alternativa descartada — Diccionario `{franja: [partidos]}`:**

La perspectiva inversa (agrupar partidos por franja) facilita calcular coincidencias, pero complica la generación de soluciones y la iteración del espacio de búsqueda. Esta vista se puede derivar fácilmente del vector cuando se necesite, por lo que no justifica usarla como estructura principal.

In [ ]:
# ---------------------------------------------------------------------------
# Definición de la representación
# ---------------------------------------------------------------------------

N_PARTIDOS = 10   # partidos por jornada
N_FRANJAS  = 10   # franjas horarias disponibles (índices 0-9)

# Mapeo índice → etiqueta legible de franja horaria
IDX_A_FRANJA = {
    0: "V20", 1: "S12", 2: "S16", 3: "S18", 4: "S20",
    5: "D12", 6: "D16", 7: "D18", 8: "D20", 9: "L20",
}

# Índices especiales para restricciones
IDX_VIERNES = 0   # V20
IDX_LUNES   = 9   # L20

# Ejemplo: jornada del enunciado
# P0=Celta-RMadrid, P1=Valencia-RSociedad, P2=Mallorca-Eibar,
# P3=Athletic-Barcelona, P4=Leganés-Osasuna, P5=Villarreal-Granada,
# P6=Alavés-Levante, P7=Espanyol-Sevilla, P8=Betis-Valladolid,
# P9=Atlético-Getafe
partidos = [
    "Celta-RMadrid", "Valencia-RSociedad", "Mallorca-Eibar",
    "Athletic-Barcelona", "Leganés-Osasuna", "Villarreal-Granada",
    "Alavés-Levante", "Espanyol-Sevilla", "Betis-Valladolid",
    "Atlético-Getafe",
]

# Solución de ejemplo (la del enunciado)
# V20, S12, S16, S18, S20, D16, D16, D18, D20, L20
solucion_ejemplo = [0, 1, 2, 3, 4, 6, 6, 7, 8, 9]

# ---------------------------------------------------------------------------
# Mostrar la asignación
# ---------------------------------------------------------------------------
print("Solución ejemplo (vector):", solucion_ejemplo)
print()
for i, franja_idx in enumerate(solucion_ejemplo):
    print(f"  P{i} ({partidos[i]:>22s}) -> franja {franja_idx} ({IDX_A_FRANJA[franja_idx]})")

# ---------------------------------------------------------------------------
# Verificación de restricciones
# ---------------------------------------------------------------------------
tiene_viernes = IDX_VIERNES in solucion_ejemplo
tiene_lunes = IDX_LUNES   in solucion_ejemplo
es_valida = tiene_viernes and tiene_lunes

print(f"\n¿Al menos 1 partido en viernes? {tiene_viernes}")
print(f"¿Al menos 1 partido en lunes?   {tiene_lunes}")
print(f"¿Solución válida?               {es_valida}")

# ---------------------------------------------------------------------------
# Coincidencias por franja (útil para penalizaciones)
# Conteo manual con vector auxiliar de tamaño k — O(n)
# ---------------------------------------------------------------------------
conteo = [0] * N_FRANJAS
for franja_idx in solucion_ejemplo:
    conteo[franja_idx] += 1

print("\nCoincidencias por franja:")
for franja_idx, cantidad in enumerate(conteo):
    if cantidad > 0:
        extra = f"  <- {cantidad - 1} coincidencia(s)" if cantidad > 1 else ""
        print(f"  {IDX_A_FRANJA[franja_idx]}: {cantidad} partido(s){extra}")

---

Según el modelo para el espacio de soluciones<br>

### (*)¿Cual es la función objetivo?

### (*)¿Es un problema de maximización o minimización?

### Respuesta

**Función objetivo:**

La función objetivo calcula la **audiencia total** (en millones de espectadores) de una jornada, sumando la contribución de cada partido. La contribución de cada partido depende de tres factores multiplicativos:

$$f(S) = \sum_{i=0}^{n-1} \text{base}_i \times \text{pond}(S_i) \times \text{penalización}(\text{coinc}_i)$$

donde, para cada partido $i$:
- $\text{base}_i$: audiencia base según las categorías de los dos equipos (en millones, referida al horario S20).
- $\text{pond}(S_i)$: coeficiente de la franja horaria asignada al partido $i$ (entre 0.4 y 1.0).
- $\text{coinc}_i = (\text{partidos en franja } S_i) - 1$: número de coincidencias en esa franja.
- $\text{penalización}(c)$: factor de reducción según la tabla de coincidencias (0 coinc → 1.0, 1 coinc → 0.75, …).

**Es un problema de maximización.** Queremos encontrar la asignación de partidos a franjas horarias que maximice la audiencia total de la jornada.

**Complejidad de evaluación:** O(n) para evaluar una solución (un recorrido para contar coincidencias por franja + un recorrido para sumar audiencias).

In [ ]:
# --- Datos del problema ---
# Audiencia base (millones) según categorías — simétrica
BASE_AUDIENCIA = {
    ("A","A"): 2.0,  ("A","B"): 1.3,  ("A","C"): 1.0,
    ("B","A"): 1.3,  ("B","B"): 0.9,  ("B","C"): 0.75,
    ("C","A"): 1.0,  ("C","B"): 0.75, ("C","C"): 0.47,
}

# Ponderación por franja horaria (índice 0-9)
PONDERACION = [0.4, 0.55, 0.7, 0.8, 1.0, 0.45, 0.75, 0.85, 1.0, 0.4]

# Factor de penalización por número de coincidencias en la misma franja
PENALIZACION = [1.0, 0.75, 0.55, 0.40, 0.30, 0.25, 0.22, 0.20, 0.20]

# Categorías de los equipos del ejemplo
categorias_ejemplo = [
    ("B","A"), ("B","A"), ("C","C"), ("B","A"), ("C","C"),
    ("B","C"), ("B","B"), ("B","B"), ("B","C"), ("B","B"),
]

def funcion_objetivo(solucion, categorias):
    """Calcula la audiencia total (en millones) de una asignación."""
    n = len(solucion)

    # Paso 1: contar partidos por franja — O(n)
    conteo = [0] * N_FRANJAS
    for franja in solucion:
        conteo[franja] += 1

    # Paso 2: sumar audiencia ponderada y penalizada — O(n)
    audiencia_total = 0.0
    for i in range(n):
        base  = BASE_AUDIENCIA[categorias[i]]
        pond  = PONDERACION[solucion[i]]
        coinc = conteo[solucion[i]] - 1          # coincidencias = partidos_en_franja - 1
        penal = PENALIZACION[coinc]
        audiencia_total += base * pond * penal

    return audiencia_total

# --- Validación con el ejemplo del enunciado ---
total = funcion_objetivo(solucion_ejemplo, categorias_ejemplo)
print(f"Audiencia total (ejemplo): {total:.4f} millones")
print(f"Valor esperado (enunciado, redondeado): 5.88 millones")

# Desglose por partido
print("\nDesglose:")
conteo = [0] * N_FRANJAS
for f in solucion_ejemplo:
    conteo[f] += 1

for i in range(N_PARTIDOS):
    base  = BASE_AUDIENCIA[categorias_ejemplo[i]]
    pond  = PONDERACION[solucion_ejemplo[i]]
    coinc = conteo[solucion_ejemplo[i]] - 1
    penal = PENALIZACION[coinc]
    valor = base * pond * penal
    print(f"  P{i} {partidos[i]:>22s}  base={base:.2f}  pond={pond:.2f}"
          f"  coinc={coinc}  penal={penal:.2f}  -> {valor:.4f}")

---

### Diseña un algoritmo para resolver el problema por fuerza bruta

### Respuesta

**Enfoque:** Enumerar **todas** las $k^n$ asignaciones posibles (donde $k = 10$ franjas y $n = 10$ partidos), verificar las restricciones para cada una, evaluar la función objetivo y quedarse con la asignación que maximice la audiencia total.

**Pseudocódigo:**

```
función FUERZA_BRUTA(categorías, n_partidos, n_franjas):
    mejor_solución <- nula
    mejor_audiencia  <- -∞
    solución ← [0, 0, ..., 0]          // primera combinación (n_partidos ceros)

    REPETIR:
        SI es_válida(solución):         // ¿tiene al menos 1 viernes y 1 lunes?
            audiencia <-> función_objetivo(solución, categorías)
            SI audiencia > mejor_audiencia:
                mejor_audiencia <- audiencia
                mejor_solución <- copia(solución)

        SI NO siguiente(solución, n_franjas):   // incrementar "odómetro" en base k
            SALIR                               // se recorrieron todas las combinaciones

    DEVOLVER mejor_solución, mejor_audiencia
```

**Generación de combinaciones - conteo en base $k$ (odómetro):**

El vector solución se trata como un número en base $k$. La función `siguiente()` incrementa la última posición y propaga el acarreo hacia la izquierda, como un odómetro:

```
[0,0,0] -> [0,0,1] -> ... → [0,0,9] -> [0,1,0] → ... → [9,9,9] -> desbordamiento (fin)
```

Esto genera las $k^n$ combinaciones sin necesidad de `itertools` ni recursión.

**Inviabilidad para $n = 10$:**

$10^{10} = 10{,}000{,}000{,}000$ combinaciones. Incluso evaluando $10^6$ por segundo, se necesitarían $\approx 10{,}000$ segundos ($\approx 2.8$ horas). Es computacionalmente inviable.

**Demostración con ejemplo reducido:**

Para verificar la corrección del algoritmo, lo ejecutamos con los primeros 6 partidos de la jornada ejemplo: $10^6 = 1{,}000{,}000$ combinaciones, factible en pocos segundos.

In [18]:
import time

def es_valida(solucion):
    """Restricciones: al menos 1 viernes (idx 0) y 1 lunes (idx 9)."""
    return IDX_VIERNES in solucion and IDX_LUNES in solucion

def siguiente(solucion, base):
    """Incrementa el vector en base k (odómetro). Devuelve False si desborda."""
    i = len(solucion) - 1
    while i >= 0:
        solucion[i] += 1
        if solucion[i] < base:
            return True
        solucion[i] = 0
        i -= 1
    return False  # desbordamiento → se recorrieron todas

def fuerza_bruta(categorias, n_partidos, n_franjas):
    """Enumera las k^n asignaciones por conteo en base k."""
    mejor_solucion = None
    mejor_audiencia = -1
    evaluadas = 0

    solucion = [0] * n_partidos      # primera combinación: [0,0,...,0]
    while True:
        if es_valida(solucion):
            evaluadas += 1
            audiencia = funcion_objetivo(solucion, categorias)
            if audiencia > mejor_audiencia:
                mejor_audiencia = audiencia
                mejor_solucion = solucion[:]
        if not siguiente(solucion, n_franjas):
            break

    return mejor_solucion, mejor_audiencia, evaluadas

# --- Demo con ejemplo reducido (6 partidos) ---
n_reducido = 6
cats_reducido = categorias_ejemplo[:n_reducido]

print(f"Fuerza bruta con {n_reducido} partidos "
      f"({N_FRANJAS}^{n_reducido} = {N_FRANJAS**n_reducido:,} combinaciones)")

t0 = time.time()
sol, aud, ev = fuerza_bruta(cats_reducido, n_reducido, N_FRANJAS)
t1 = time.time()

print(f"Soluciones válidas evaluadas: {ev:,}")
print(f"Tiempo: {t1-t0:.2f} s")
print(f"\nMejor audiencia: {aud:.4f} millones")
print(f"Mejor solución:  {sol}")
for i in range(n_reducido):
    print(f"  P{i} ({partidos[i]:>22s}) -> {IDX_A_FRANJA[sol[i]]}")

# --- Estimación para n=10 ---
factor = N_FRANJAS ** (N_PARTIDOS - n_reducido)
t_estimado = (t1 - t0) * factor
print(f"\nPara n={N_PARTIDOS}: {N_FRANJAS}^{N_PARTIDOS} = "
      f"{N_FRANJAS**N_PARTIDOS:,} combinaciones")
print(f"Tiempo estimado: {t_estimado:.0f} s "
      f"({t_estimado/3600:.1f} h) -> INVIABLE")

Fuerza bruta con 6 partidos (10^6 = 1,000,000 combinaciones)
Soluciones válidas evaluadas: 199,262
Tiempo: 0.31 s

Mejor audiencia: 4.6810 millones
Mejor solución:  [4, 7, 0, 8, 9, 3]
  P0 (         Celta-RMadrid) -> S20
  P1 (    Valencia-RSociedad) -> D18
  P2 (        Mallorca-Eibar) -> V20
  P3 (    Athletic-Barcelona) -> D20
  P4 (       Leganés-Osasuna) -> L20
  P5 (    Villarreal-Granada) -> S18

Para n=10: 10^10 = 10,000,000,000 combinaciones
Tiempo estimado: 3102 s (0.9 h) -> INVIABLE


---

### Calcula la complejidad del algoritmo por fuerza bruta

### Respuesta

El algoritmo de fuerza bruta tiene tres operaciones principales que se ejecutan para cada una de las $k^n$ combinaciones:

| Operación | Coste por llamada | Llamadas |
|:--|:--:|:--:|
| `siguiente()`: incrementar odómetro | $O(n)$ peor caso, $O(1)$ amortizado | $k^n$ |
| `es_valida()`: buscar viernes y lunes en el vector | $O(n)$ | $k^n$ |
| `funcion_objetivo()`: contar coincidencias + sumar audiencias | $O(n)$ | $\leq k^n$ (solo válidas) |

**Complejidad temporal:**

Cada iteración del bucle realiza trabajo $O(n)$ (validación + evaluación + avance del odómetro), y hay $k^n$ iteraciones:

$$T(n, k) = O(n \cdot k^n)$$

Con los valores del problema ($k = n = 10$):

$$T(10, 10) = O(10 \cdot 10^{10}) = O(10^{11})$$

La complejidad es **exponencial** en $n$: El factor dominante es $k^n$, que crece mucho más rápido que el factor lineal $n$.

**Complejidad espacial:**

$$S(n, k) = O(n + k)$$

- $O(n)$: el vector solución actual + la mejor solución guardada.
- $O(k)$: el vector auxiliar de conteo por franja dentro de `funcion_objetivo`.
- No se almacenan todas las combinaciones en memoria; se generan una a una (generación implícita).

**Clase de complejidad:**

El algoritmo pertenece a la clase $O(k^n)$ tiempo exponencial. Es intratable para valores moderados de $n$ y $k$, lo que motiva la búsqueda de algoritmos más eficientes (greedy, backtracking, metaheurísticas).

In [19]:
# --- Complejidad del algoritmo de fuerza bruta ---
# T(n,k) = O(n · k^n)   |   S(n,k) = O(n + k)

print("Crecimiento de la complejidad temporal T(n,k) = n · k^n")
print(f"{'n':>3s} {'k':>3s} {'k^n':>18s} {'n·k^n':>20s}")
print("-" * 48)

k = N_FRANJAS
for n in range(1, N_PARTIDOS + 1):
    comb = k ** n
    ops  = n * comb
    print(f"{n:>3d} {k:>3d} {comb:>18,} {ops:>20,}")

# Verificación empírica con los tiempos medidos en cell-23
print(f"\n--- Verificación empírica (cell-23) ---")
t_medido = t1 - t0  # tiempo medido para n=6
ops_n6 = n_reducido * (N_FRANJAS ** n_reducido)
ops_n10 = N_PARTIDOS * (N_FRANJAS ** N_PARTIDOS)
print(f"n=6:  {ops_n6:,} operaciones → {t_medido:.2f} s medidos")
print(f"n=10: {ops_n10:,} operaciones → factor ×{ops_n10/ops_n6:,.0f} respecto a n=6")
print(f"      Tiempo estimado: {t_medido * (ops_n10/ops_n6):.0f} s "
      f"({t_medido * (ops_n10/ops_n6) / 3600:.1f} h)")

Crecimiento de la complejidad temporal T(n,k) = n · k^n
  n   k                k^n                n·k^n
------------------------------------------------
  1  10                 10                   10
  2  10                100                  200
  3  10              1,000                3,000
  4  10             10,000               40,000
  5  10            100,000              500,000
  6  10          1,000,000            6,000,000
  7  10         10,000,000           70,000,000
  8  10        100,000,000          800,000,000
  9  10      1,000,000,000        9,000,000,000
 10  10     10,000,000,000      100,000,000,000

--- Verificación empírica (cell-23) ---
n=6:  6,000,000 operaciones → 0.31 s medidos
n=10: 100,000,000,000 operaciones → factor ×16,667 respecto a n=6
      Tiempo estimado: 5170 s (1.4 h)


---

(*)Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta

Respuesta

(*)Calcula la complejidad del algoritmo

Respuesta

Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios

Respuesta

Aplica el algoritmo al juego de datos generado

Respuesta

Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo

Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño

Respuesta

## Referencias

**P1 - Principio multiplicativo (posibilidades sin restricciones):**

1. Reyero Díez, R. *"Algoritmos de Optimización"*, Manual de la asignatura, VIU. Cap. 1 - Complejidad computacional.
2. VIU-03MIAR — Sesión 02 — VC1 - Introducción a los Algoritmos (conteo de posibilidades, fuerza bruta).

**P2 - Principio de inclusión-exclusión (posibilidades con restricciones):**

3. Wikipedia: *Principio de inclusión-exclusión*. https://es.wikipedia.org/wiki/Principio_de_inclusión-exclusión

**P3 - Estructura de datos para el espacio de soluciones:**

4. VIU-03MIAR — AG1: Problema de las N-Reinas. Representación vectorial: posición = entidad, valor = asignación. Precedente directo de la estructura elegida para el problema de La Liga.